keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,820 @@
+- The procedure of establishing new nodes is considerably more time-intensive than expanding the capabilities of current nodes.
+- Improved inter-pod network latencies as the inter-node traffic can be reduced if more pods can be hosted on a single node.
+- Easier to manage the cluster with fewer nodes, which brings less overhead on the control-plane
+- Mitigate a few of the existing limitations/issues that are associated with a node/kubelet restart.
+
+Implementing this KEP will empower nodes to recognize and adapt to changes in their compute configurations and allow facilitating the efficient and effective deployment of pod workloads to nodes capable of meeting the required compute demands.
+
+### Goals
+
+* Achieve seamless node capacity expansion through hot plugging resources.
+* Enable the re-initialization of resource managers (CPU manager, memory manager) to accommodate alterations in the node's resource allocation.
+* Recalculating the OOMScoreAdj and swap memory limit for existing pods.
+
+### Non-Goals
+
+* Dynamically adjust system reserved and kube reserved values.
+* Hot unplug of node resources.
+* Update the autoscaler to utilize resource hot plugging.
+* Re-balance workloads across the nodes.
+* Update runtime/NRI plugins with host resource changes.
+
+## Proposal
+
+This KEP strives to enable node resource hot plugging by incorporating a polling mechanism within the kubelet to retrieve machine-information from cAdvisor's cache, which is already updated periodically.
+The kubelet will periodically fetch this information, subsequently entrusting the node status updater to disseminate these updates at the node level across the cluster.
+Moreover, this KEP aims to refine the initialization and reinitialization processes of resource managers, including the memory manager and CPU manager, to ensure their adaptability to changes in node configurations.
+With this proposal its also necessary to recalculate and update OOMScoreAdj and swap limit for the pods that had been existing before resize and tradeoff between performance overhead and OOMScoreAdj and swap limit correctness.
+
+Update in OOMScoreAdjust and swap limit
+    - The recalculation will help the existing pods to be closer to the expected value post-resize but carries an inherent overhead on the kubelet.
+
+Not Updating OOMScoreAdjust and swap limit
+    - Efficient, We can avoid the performance overhead required for recalculating the scores and may lead to incorrect eviction, with new pods being evicted as OOMScoreAdj will be high.
+
+### User Stories
+
+#### Story 1
+
+As a Kubernetes user, I want to allocate more resources (CPU, memory) to a node with existing specialized hardware or CPU Capabilities (for example:https://www.kernel.org/doc/html/v5.8/arm64/elf_hwcaps.html)
+so that additional workloads can leverage the hardware to be efficiently scheduled and run without manual intervention.
```

Comment:
Is there a non-functional requirement that you want to do this without a reboot?

Reply:
Yes, we would want to update the node capacity in a streamlined fashion without causing any disruption to any existing workloads.